### PR TITLE
Several Gradient improvements (more interpolation modes, gradient reverse button, ...)

### PR DESCRIPTION
**Pull request by Geometror**
This PR aims to improve the Gradient resource by providing more interpolation modes, a gradient reverse button and restructuring the properties.
Implements godotengine/godot-proposals#1933 and godotengine/godot-proposals#1724

main

CHANGES:
- Add 2 more interpolation modes to the Gradient resource:
	- Constant
	- Cubic

	grafik
- Add reverse gradient function and the corresponding button to the inspector plugin 
- Restructured/cleaned up gradient properties (offsets and colors which are used rarely are now in a collapsed group called "raw data"
- Use GradientTexture for the preview of the inspector gradient editor (reduce code and gain flexibility, no visual difference)
- Fix gradient inspector plugin not updating when properties are changed

CURRENT MASTER:
grafik

THIS PR:
grafik

Note:
I know, a custom button is a bit overkill but I ran into a problem when using the normal button because the icon stayed invisible when it was set with `btn->set_icon(get_theme_icon(...))` in `EditorInspectorPlugin::parse_begin()` [only if an svg icon is used, which is rendered at startup, so I guess the svg icons weren't rasterized yet] Furthermore, I couldn't find an easy way to update the icon of the button after a theme change which didn't involve lots of passthrough-functions.
If someone has an idea how to solve this with a general button let me know, but for now I think its fine (and even more light-weight since there is no TextParagraph instantiated).

As always, feedback is welcome :)

### CODE REVIEW

**Comment by Calinou** [comment]
I'd rename the "constant" mode to "nearest" to be consistent with the texture filtering mode's name.

**Comment by Geometror** [comment]
Nearest neighbor would mean that for every sampled point the color of the nearest gradient point would be returned, but that is not the case as the color of one segment is determined by the left of two points and stays constant between points.
In Blender and other software it is called constant interpolation, so I went with that, but the term "stairs" is also sometimes used.

grafik
I could add nearest neighbor as well, but I think that constant/stair interpolation is enough (nearest isn't really that useful here).

File: editor/plugins/gradient_editor_plugin.cpp [hunk 1]
```diff
@@ -77,6 +89,24 @@ GradientEditor::GradientEditor() {
...
 ///////////////////////
 
+void GradientReverseButton::_notification(int p_what) {
+	if (p_what == NOTIFICATION_DRAW) {
+		Ref<Texture2D> icon = get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons"));
+		if (is_pressed()) {
+			draw_texture_rect(icon, Rect2(margin * EDSCALE, margin * EDSCALE, icon->get_width() * EDSCALE, icon->get_height() * EDSCALE), false, Color(0.8, 0.8, 0.8));
+
+		} else {
+			draw_texture_rect(icon, Rect2(margin * EDSCALE, margin * EDSCALE, icon->get_width() * EDSCALE, icon->get_height() * EDSCALE));
```
Comment (bruvzg) [inline editor/plugins/gradient_editor_plugin.cpp:99]:
```suggestion
			draw_texture_rect(icon, Rect2(margin, margin, icon->get_width(), icon->get_height()), false, Color(0.8, 0.8, 0.8));

		} else {
			draw_texture_rect(icon, Rect2(margin, margin, icon->get_width(), icon->get_height()));
```

File: editor/plugins/gradient_editor_plugin.cpp [hunk 2]
```diff
@@ -77,6 +89,24 @@ GradientEditor::GradientEditor() {
...
+		}
+	}
+}
+
+Size2 GradientReverseButton::get_minimum_size() const {
+	return (get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons"))->get_size() + Size2(margin * 2, margin * 2)) * EDSCALE;
```
Comment (bruvzg) [inline editor/plugins/gradient_editor_plugin.cpp:105]:
```suggestion
	return (get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons"))->get_size() + Size2(margin * 2, margin * 2));
```

**Review by bruvzg — COMMENTED** [review]
Reverse icon in the hiDPI mode is too large (`EDSCALE` should not be applied here, it's  already taken into account by the theme icon loader).

Screenshot 2021-10-04 at 10 23 44

Everything else seems to be fine.

**Comment by Geometror** [comment]
Fixed and rebased.

**Review by bruvzg — APPROVED** [review]

File: scene/gui/gradient_edit.h [hunk 3]
```diff
@@ -45,6 +45,10 @@ class GradientEdit : public Control {
 	bool grabbing = false;
 	int grabbed = -1;
 	Vector<Gradient::Point> points;
+	Gradient::InterpolationMode interpolation_mode = Gradient::GRADIENT_INTERPOLATE_LINEAR;
+
+	Ref<Gradient> gradient_cache;
+	Ref<GradientTexture> preview_texture;
```
Comment (Chaosus) [inline scene/gui/gradient_edit.h:51]: Since https://github.com/godotengine/godot/pull/54715 was merged I think it should be renamed to `GradientTexture1D`

Comment (Geometror) [reply scene/gui/gradient_edit.h]: Rebased.

**Comment by Chaosus** [comment]
I think it's fine, thanks!

**Pull request merged**